src/runtime/test_case_runner.js

```diff
@@ -166,26 +172,40 @@ export default class TestCaseRunner {
 
   async run() {
     this.emitPrepared()
-    this.emit('test-case-started', {})
-    await this.runHooks(
-      this.beforeHookDefinitions,
-      {
-        sourceLocation: this.testCaseSourceLocation,
-        pickle: this.testCase.pickle,
-      },
-      true
-    )
-    await this.runSteps()
-    await this.runHooks(
-      this.afterHookDefinitions,
-      {
-        sourceLocation: this.testCaseSourceLocation,
-        pickle: this.testCase.pickle,
-        result: this.result,
-      },
-      false
-    )
-    this.emit('test-case-finished', { result: this.result })
+    let retries = 0
+    for (; retries <= this.retries; retries++) {
+      this.emit('test-case-started', {})
+      if (retries > 0) {
+        this.resetTestProgressData()
+        this.result.retryAttempt = retries
+      }
+      await this.runHooks(
+        this.beforeHookDefinitions,
+        {
+          sourceLocation: this.testCaseSourceLocation,
+          pickle: this.testCase.pickle,
+        },
+        true
+      )
+      await this.runSteps()
+      await this.runHooks(
+        this.afterHookDefinitions,
+        {
+          sourceLocation: this.testCaseSourceLocation,
+          pickle: this.testCase.pickle,
+          result: this.result,
+        },
+        false
+      )
+      const shouldRetryTestCase =
+        this.result.status === Status.FAILED && retries < this.retries
+      if (!shouldRetryTestCase) {
+        this.emit('test-case-finished', { result: this.result })
+        break
+      }
+      this.result.status = Status.FLAKY
+      this.emit('test-case-finished', { result: this.result })
```

Comment:
Thoughts on rearranging this a bit and bringing the reset of the test progress data down here? 
```js
const shouldRetry = // ...
if (shouldRetry) {
  // update result status
}
// emit result
if (shouldRetry) {
  // reset progress data 
} else {
  break
}
```

Reply:
I've grouped all the `shouldRetry` logic at the bottom of the loop